rustbus/src/connection/ll_conn.rs

```diff
@@ -56,12 +57,11 @@ impl RecvConn {
     /// Reads from the source once but takes care that the internal buffer only reaches at maximum max_buffer_size
     /// so we can process messages separatly and avoid leaking file descriptors to wrong messages
     fn refill_buffer(&mut self, max_buffer_size: usize, timeout: Timeout) -> Result<()> {
-        let bytes_to_read = max_buffer_size - self.msg_buf_in.len();
-
-        const BUFSIZE: usize = 512;
-        let mut tmpbuf = [0u8; BUFSIZE];
+        if self.msg_buf_in.len() < max_buffer_size {
```

Comment:
I think this needs to be `if self.msg_buf_in.len() != max_buffer_size {` so that `msg_buf_in` also gets resized to be smaller if the next message is smaller than the previous one.

Reply:
Why? I specifically didn't do this to avoid unnecessary `memset`s.

Reply:
Because the `msg_buf_in` should only ever contain as many bytes as the message that is currently being read

Reply:
This information is stored in `msg_buf_filled`. Imagine you receive a sequence of messages with the following sizes: `1K`, `128B`, `512B`, `1K`. The way this PR works right now, resizes would write `1K` of zeroes. With your suggestion it would write at least `2K-128B` of zeroes (best case, if each message is read in one recvmsg).

Reply:
That is true but I would accept that cost for the sake of an easier invariant and making the `msg_buf_filled` field unnecessary. The few memsets are unlikely to be a bottleneck in the library.

Another thing I don't really like is that the code currently "reuses" the `msg_buf_in` by cloning (part of) it when a full message has been received. I cases where there are typically small messages received one big message will forever bloat the memory usage.

Reply:
> Could you point where the cloning is?

[`rustbus/src/wire/unmarshal.rs:175`](https://github.com/KillingSpark/rustbus/blob/28127b3d935654e2c76a88e4b2d5b7f1dff08bfb/rustbus/src/wire/unmarshal.rs#L175)

I think it would be better to do a swap here with a newly allocated vec of a sensible default size instead of doing it like this.

> Right. Exactly as it was. .clear() does not release any memory

Yes, I am aware. I wanted to say that it's currently unsatisfying and that, while we are working on this anyways we could improve this aspect too.

Reply:
> > Could you point where the cloning is?
> 
> [`rustbus/src/wire/unmarshal.rs:175`](https://github.com/KillingSpark/rustbus/blob/28127b3d935654e2c76a88e4b2d5b7f1dff08bfb/rustbus/src/wire/unmarshal.rs#L175)
> 
> I think it would be better to do a swap here with a newly allocated vec of a sensible default size instead of doing it like this. Or maybe there is an even better way?

Thanks! I completely missed this. A swap, even with an empty vector, sounds reasonable.

Reply:
So, should I `.drain(..offset)` bytes from the buffer, or keep the offset to avoid shifting?

Reply:
I think draining is fine as a first step towards improvement. Ideally the initial offset would be passed through to the unmarshalling code though. But that could be a separate PR, or you could leave it for me :)

Reply:
I've implemented the take+drain part, but left the `msg_buf_filled` for now. The annoying thing is that without it the message must be resized back in case of any error.